bluepysnap/network.py

```diff
@@ -0,0 +1,170 @@
+
+    def keys(self):
+        """Returns iterator on the NetworkObjectPopulation names.
+
+        Made to simulate the behavior of a dict.keys().
+        """
+        return (name for name in self.population_names)
+
+    def values(self):
+        """Returns iterator on the NetworkObjectPopulations.
+
+        Made to simulate the behavior of a dict.values().
+        """
+        return (self[name] for name in self.population_names)
+
+    def items(self):
+        """Returns iterator on the tuples (population name, NetworkObjectPopulations).
+
+        Made to simulate the behavior of a dict.items().
+        """
+        return ((name, self[name]) for name in self.population_names)
+
+    def __getitem__(self, population_name):
+        """Access the NetworkObjectPopulation corresponding to the population 'population_name'."""
+        try:
+            return self._populations[population_name]
+        except KeyError:
+            raise BluepySnapError("{} not a {} population.".format(population_name, self.__class__))
+
+    def __iter__(self):
+        """Allows iteration over the different NetworkObjectPopulation."""
+        return iter(self.keys())
+
+    @cached_property
+    def size(self):
+        """Total number of NetworkObject inside the circuit."""
+        return sum(pop.size for pop in self.values())
+
+    @cached_property
+    def property_names(self):
+        """Returns all the NetworkObject properties present inside the circuit."""
+        return set(prop for pop in self.values() for prop in pop.property_names)
+
+    def _get_ids_from_pop(self, fun_to_apply, returned_ids_cls):
+        """Get CircuitIds of class 'returned_ids_cls' for all populations using 'fun_to_apply'.
+
+        Args:
+            fun_to_apply (function): A function that returns the list of IDs for each population
+                and the population containing these IDs.
+            returned_ids_cls (CircuitNodeIds/CircuitEdgeIds): the class for the CircuitIds.
+
+        Returns:
+            CircuitNodeIds/CircuitEdgeIds: containing the IDs and the populations.
+        """
+        str_type = "<U{}".format(max(len(pop) for pop in self.population_names))
+        ids = []
+        populations = []
+        for pop in self.values():
+            pop_ids, name_ids = fun_to_apply(pop)
+            pops = np.full_like(pop_ids, fill_value=name_ids, dtype=str_type)
+            ids.append(pop_ids)
+            populations.append(pops)
+        ids = np.concatenate(ids).astype(np.int64)
+        populations = np.concatenate(populations).astype(str_type)
+        return returned_ids_cls.from_arrays(populations, ids)
+
+    @abc.abstractmethod
+    def ids(self, *args, **kwargs):
+        """Resolves the ids of the NetworkObject."""
+
+    @abc.abstractmethod
+    def get(self, group=None, properties=None):
+        """Returns the properties of a the NetworkObject."""
+        ids = self.ids(group)
+        properties = utils.ensure_list(properties)
```

Comment:
you could have a properties_set var here since you do set(properties) multiple times.

Reply:
indeed